etna/reconciliation/top_down.py

```diff
@@ -0,0 +1,132 @@
+        -------
+        :
+            Fitted instance of reconciliator.
+        """
+        if ts.hierarchical_structure is None or ts.current_df_level is None:
+            raise ValueError(f"The method can be applied only to instances with a hierarchy!")
+
+        current_level_index = ts.hierarchical_structure.get_level_depth(ts.current_df_level)
+        source_level_index = ts.hierarchical_structure.get_level_depth(self.source_level)
+        target_level_index = ts.hierarchical_structure.get_level_depth(self.target_level)
+
+        if target_level_index < source_level_index:
+            raise ValueError("Target level should be lower or equal in the hierarchy than the source level!")
+
+        if current_level_index < target_level_index:
+            raise ValueError("Current TSDataset level should be lower or equal in the hierarchy than the target level!")
+
+        source_level_ts = ts.get_level_dataset(self.source_level)
+        target_level_ts = ts.get_level_dataset(self.target_level)
+
+        if source_level_ts.hierarchical_structure is None:
+            raise ValueError("Source level dataset has no hierarchical structure!")
+
+        if target_level_ts.hierarchical_structure is None:
+            raise ValueError("Target level dataset has no hierarchical structure!")
+
+        if source_level_index < target_level_index:
+
+            summing_matrix = target_level_ts.hierarchical_structure.get_summing_matrix(
+                target_level=self.source_level, source_level=self.target_level
+            )
+
+            source_level_segments = source_level_ts.hierarchical_structure.get_level_segments(self.source_level)
+            target_level_segments = target_level_ts.hierarchical_structure.get_level_segments(self.target_level)
+
+            self.mapping_matrix = lil_matrix((len(target_level_segments), len(source_level_segments)))
+
+            for source_index, target_index in zip(*summing_matrix.nonzero()):
+                source_segment = source_level_segments[source_index]
+                target_segment = target_level_segments[target_index]
+
+                self.mapping_matrix[target_index, source_index] = self.proportions_method(
+                    target_series=target_level_ts[:, target_segment, "target"],
+                    source_series=source_level_ts[:, source_segment, "target"],
+                )
+
+        else:
+            self.mapping_matrix = target_level_ts.hierarchical_structure.get_summing_matrix(
+                target_level=self.target_level, source_level=self.source_level
+            )
+
+        return self
+
+    def _estimate_ahp_proportion(self, target_series: pd.Series, source_series: pd.Series) -> float:
+        """Calculate reconciliation proportion with Average historical proportions method."""
+        data = pd.concat((target_series, source_series), axis=1).values
+        data = data[-self.period_length :]
+        return np.nanmean(data[..., 0] / data[..., 1])
```

Comment:
May be we can you bottleneck library here like in statistics [see](https://github.com/tinkoff-ai/etna/blob/f46b363b9fffe8d0e5f52992fd655e69dcaf8683/etna/transforms/math/statistics.py#L177)